File: hackerrank/problem solving/algorithms/2 - implementation/51 - service lane/service_lane.rs
// https://www.hackerrank.com/challenges/service-lane/problem?isFullScreen=true

fn main() {
    let n: i32 = text_io::read!();
    let t: i32 = text_io::read!();
    let widths: Vec<i32> = read_nums(n);
    for _ in 0..t {
        println!("{}", min_width_in_segment(&widths, &read_nums(2)))
    }
}

fn read_nums(n: i32) -> Vec<i32> {
    (0..n).map(|_| text_io::read!()).collect()
}

// n: length of widths
// T: O(n)
// S: O(1) extra space
fn min_width_in_segment(widths: &[i32], segment: &[i32]) -> i32 {
    let start = segment[0] as usize;
    let finish = segment[1] as usize;
    *widths[start..=finish]
        .iter()
        .min()
        .unwrap()
}

File: hackerrank/problem solving/algorithms/2 - implementation/51 - service lane/service_lane.ts
// https://www.hackerrank.com/challenges/service-lane/problem?isFullScreen=true

function main() {
  const [_, t]: number[] = readNums();
  const widths = readNums();
  for (let i = 0; i < t; i++)
    console.log(minWidthInSegment(widths, readNums()));
}

function readNums(): number[] {
  return readLine().split(' ').map(Number);
}

// n: length of widths
// T: O(n)
// S: O(1) extra space
function minWidthInSegment(widths: number[], segment: number[]): number {
  const [start, finish]: number[] = segment;
  let minEl: number = widths[start];
  for (let i = start; i <= finish; ++i) minEl = Math.min(minEl, widths[i]);
  return minEl;
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
const inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}
